internal/services/controllers/workflows/queue.go

```diff
@@ -538,90 +539,93 @@ func (ec *WorkflowsControllerImpl) runGetGroupKeyRunReassignTenant(ctx context.C
 	return g.Wait()
 }
 
+func (wc *WorkflowsControllerImpl) getLock(key string) *sync.Mutex {
+	actual, _ := wc.queueMutex.LoadOrStore(key, &sync.Mutex{})
+	return actual.(*sync.Mutex)
+}
+
 func (wc *WorkflowsControllerImpl) queueByCancelInProgress(ctx context.Context, tenantId, groupKey string, workflowVersion *dbsqlc.GetWorkflowVersionForEngineRow) error {
 	ctx, span := telemetry.NewSpan(ctx, "queue-by-cancel-in-progress")
 	defer span.End()
 
 	wc.l.Info().Msgf("handling queue with strategy CANCEL_IN_PROGRESS for %s", groupKey)
 
-	// list all workflow runs that are running for this group key
-	running := db.WorkflowRunStatusRunning
+	mutex := wc.getLock(fmt.Sprintf("%s:%s", tenantId, groupKey))
+	mutex.Lock()
```

Comment:
```suggestion
	if ok := mutex.TryLock(); !ok {
	  return nil 
        }
```

Perhaps we should just exit if we can't get a lock, otherwise we'll be blocked on this method for a while if we get a lot of workflow runs in.